File: packages/soulseek_protocol/lib/src/services/wishlist_service.dart
import 'dart:async';

import '../connection/server_connection.dart';
import '../messages/buffer.dart';
import '../messages/codes.dart';
import '../messages/message.dart';

class WishlistService {
  final ServerTransport _server;
  final _resultsController = StreamController<WishlistReply>.broadcast();
  int _nextTicket = 1;

  late final StreamSubscription _messageSub;

  WishlistService({required ServerTransport server}) : _server = server;

  Stream<WishlistReply> get wishlistResults => _resultsController.stream;

  void init() {
    _messageSub = _server.messages.listen(_onMessage);
  }

  int wishlistSearch(String query) {
    final ticket = _nextTicket++;
    _server.sendMessage(SoulseekMessage(
      ServerCode.wishes,
      WishlistSearchRequest(ticket: ticket, query: query).serialize().toBytes(),
    ));
    return ticket;
  }

  void addWishlistItem(String phrase) {
    _server.sendMessage(SoulseekMessage(
      ServerCode.wishlistInclusion,
      WishlistInclusion(add: true, phrase: phrase).serialize().toBytes(),
    ));
  }

  void removeWishlistItem(String phrase) {
    _server.sendMessage(SoulseekMessage(
      ServerCode.wishlistInclusion,
      WishlistInclusion(add: false, phrase: phrase).serialize().toBytes(),
    ));
  }

  void _onMessage(SoulseekMessage message) {
    if (message.code != ServerCode.wishReply) return;

    try {
      final reply = WishlistReply.parse(ReadBuffer(message.payload));
      _resultsController.add(reply);
    } catch (_) {
      // Skip malformed wishlist replies
    }
  }

  void dispose() {
    _messageSub.cancel();
    _resultsController.close();
  }
}

File: Soulseek.Protocol/Services/WishlistService.cs
using Soulseek.Protocol.Connection;
using Soulseek.Protocol.Messages;

namespace Soulseek.Protocol.Services;

public class WishlistService
{
    private readonly IServerTransport _server;
    private readonly Subject<WishlistReply> _resultsSubject = new();
    private int _nextTicket = 1;
    private IDisposable? _messageSub;

    public WishlistService(IServerTransport server)
    {
        _server = server;
    }

    public IObservable<WishlistReply> WishlistResults => _resultsSubject;

    public void Init()
    {
        _messageSub = _server.MessageStream.Subscribe(OnMessage);
    }

    public int WishlistSearch(string query)
    {
        var ticket = _nextTicket++;
        _server.SendMessage(new SoulseekMessage(
            ServerCode.Wishes,
            new WishlistSearchRequest(ticket, query).Serialize().ToBytes()
        ));
        return ticket;
    }

    public void AddWishlistItem(string phrase)
    {
        _server.SendMessage(new SoulseekMessage(
            ServerCode.WishlistInclusion,
            new WishlistInclusion(true, phrase).Serialize().ToBytes()
        ));
    }

    public void RemoveWishlistItem(string phrase)
    {
        _server.SendMessage(new SoulseekMessage(
            ServerCode.WishlistInclusion,
            new WishlistInclusion(false, phrase).Serialize().ToBytes()
        ));
    }

    private void OnMessage(SoulseekMessage message)
    {
        if (message.Code != ServerCode.WishReply) return;

        try
        {
            var reply = WishlistReply.Parse(new ReadBuffer(message.Payload));
            _resultsSubject.OnNext(reply);
        }
        catch
        {
            // Skip malformed wishlist replies
        }
    }

    public void Dispose()
    {
        _messageSub?.Dispose();
    }
}